velox/exec/AggregateWindow.cpp

```diff
@@ -179,24 +183,56 @@ class AggregateWindowFunction : public exec::WindowFunction {
     bool usePreviousAggregate;
   };
 
+  bool handleAllEmptyFrames(
+      const SelectivityVector& validRows,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    if (!validRows.hasSelections()) {
+      uint64_t* rawNulls = result->mutableRawNulls();
+      for (auto row = 0; row < validRows.size(); row++) {
+        bits::setNull(rawNulls, row + resultOffset, true);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  // Computes the least frameStart row and the max frameEnds row
+  // indices for the valid frames of this output block. These indices are used
+  // as bounds when reading input parameter vectors for aggregation.
+  // This method expects to have at least 1 valid frame in the block.
+  // Blocks with all empty frames are handled before this point.
   FrameMetadata analyzeFrameValues(
+      const SelectivityVector& validRows,
       const vector_size_t* rawFrameStarts,
-      const vector_size_t* rawFrameEnds,
-      vector_size_t numRows) {
-    vector_size_t firstRow = rawFrameStarts[0];
+      const vector_size_t* rawFrameEnds) {
+    VELOX_DCHECK(validRows.hasSelections());
+    // Copy validRows into a modifiable SelectivityVector.
+    validRowsWriteCopy_.resize(validRows.size(), false);
+    validRowsWriteCopy_.select(validRows);
+
+    // Use first valid frame row for the initialization.
+    auto firstValidRow = validRowsWriteCopy_.begin();
+    vector_size_t firstRow = rawFrameStarts[firstValidRow];
     vector_size_t fixedFrameStartRow = firstRow;
-    vector_size_t lastRow = rawFrameEnds[0];
+    vector_size_t lastRow = rawFrameEnds[firstValidRow];
+    vector_size_t prevFrameEnds = lastRow;
+
+    // Continue iterating on the rest of the frames.
+    validRowsWriteCopy_.setActiveRange(firstValidRow + 1, validRows.end());
 
     bool incrementalAggregation = true;
-    for (int i = 1; i < numRows; i++) {
+    validRowsWriteCopy_.applyToSelected([&](auto i) {
```

Comment:
Why can't we just use 'validRows' here?

Reply:
If we used validRows I would need to check an if condition for the row Index (as we want to skip the first row in this loop) in each iteration. This could be wasteful. And I can't modify validRows as it is a const&. So in general using this copy seemed like a reasonable compromise.

Reply:
Do we have to skip first row? We didn't skip it before. I might be missing something and would appreciate a clarification.

Reply:
Previously the first row was skipped as we iterated from index 1 to numRows. But on some more thought (and maybe that was your intent already) we needn't skip the first row in the processing. Even if the loop starts from the first index, it logic will just retain the original initialized values.

The new revised code is cleaner and removes the member variable.

Reply:
That's great. Thank you for revising.